src/main/java/com/google/sps/servlets/BlobServlet.java

```diff
@@ -33,59 +41,203 @@
   public void doPost(HttpServletRequest request, HttpServletResponse response)
       throws IOException {
     UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
+    // Must be logged in
     if (!userService.isUserLoggedIn()) {
-      response.sendRedirect("/imgupload.html");
+      throw new IOException("User must be logged in.");
+    }
+
+    // Mode is a required parameter
+    String mode = request.getParameter("mode");
+    if (mode == null || (mode != "create" && mode != "update")) {
+      throw new IOException("Invalid mode.");
+    }
+
+    // Check if working with image or mask
+    String parentImg = request.getParameter("parent-img");
+    Boolean isMask = false;
+    if (parentImg != null) {
+      isMask = true;
+    }
+
+    // Get the image name entered by the user
+    String imgName = request.getParameter("img-name");
+    if (imgName == null) {
+      throw new IOException("Image name must be provided.");
+    }
+
+    String projId = request.getParameter("proj-id");
+    Key projKey = KeyFactory.stringToKey(projId);
+    Entity projEntity = new Entity("Project");
+    try {
+      projEntity = datastore.get(projKey);
+    } catch (Exception e) {
+      response.sendRedirect("/");
       return;
     }
+    String uid = userService.getCurrentUser().getUserId();
+    Key assetParentKey = projKey;
+    Entity imgEntity = new Entity("Image", projKey);
 
-    // Get the name entered by the user
-    String name = request.getParameter("name");
+    // Must be an owner or editor
+    Query projQuery = new Query("Project");
+
+    Filter ownEditFilter = new CompositeFilter(
+        CompositeFilterOperator.AND,
+        Arrays.asList(
+            new FilterPredicate("proj-id", FilterOperator.EQUAL, projId),
+            new CompositeFilter(
+                CompositeFilterOperator.OR,
+                Arrays.<Filter>asList(
+                    new FilterPredicate("owners", FilterOperator.EQUAL, uid),
+                    new FilterPredicate("editors", FilterOperator.EQUAL,
+                                        uid)))));
+
+    projQuery.setFilter(ownEditFilter);
+    PreparedQuery accessibleProjects = datastore.prepare(projQuery);
+
+    // Only owners and editors of given project can modify or create
+    if (accessibleProjects.countEntities() == 0) {
+      throw new IOException(
+          "User does not have permission to edit the project.");
+    }
+
+    // Asset to update must already exist
+    if (isMask) {
+      Query imgQuery = new Query("Image").setAncestor(projKey);
+      Filter imgFilter =
+          new FilterPredicate("name", FilterOperator.EQUAL, parentImg);
+      imgQuery.setFilter(imgFilter);
+      PreparedQuery existingImg = datastore.prepare(imgQuery);
```

Comment:
This is repeated below so it would be good to break it out into a helper function.

Reply:
Noted